src/engine/BMInterfaceTournament.php

```diff
@@ -0,0 +1,978 @@
+                            ':tournament_state' => $tournament->tournamentState,
+                            ':round_number' => $tournament->roundNumber,
+                            ':tournament_id' => $tournament->tournamentId);
+        self::$db->update($query, $parameters);
+    }
+
+    /**
+     * Logic to save player parameters
+     *
+     * @param BMTournament $tournament
+     */
+    protected function save_player_parameters(BMTournament $tournament) {
+        foreach ($tournament->playerIdArray as $playerIdx => $playerId) {
+            $query = 'UPDATE tournament_player_map '.
+                     'SET position = :position,'.
+                     '    remain_count = :remain_count '.
+                     'WHERE tournament_id = :tournament_id '.
+                     'AND player_id = :player_id;';
+            $parameters = array(':position' => $playerIdx + 1,
+                                ':remain_count' => $tournament->remainCountArray[$playerIdx],
+                                ':tournament_id' => $tournament-> tournamentId,
+                                ':player_id' => $playerId);
+            self::$db->update($query, $parameters);
+        }
+    }
+
+    /**
+     * Force a tournament to update
+     *
+     * @param int $tournamentId
+     */
+    protected function update_tournament($tournamentId) {
+        if ($this->isTest) {
+            return;
+        }
+
+        $tournament = $this->load_tournament($tournamentId);
+        $this->save_tournament($tournament);
+    }
+
+    /**
+     * Retrieve the status of a tournament
+     *
+     * @param BMTournament $tournament
+     * @return string
+     */
+    protected function get_tournament_status(BMTournament $tournament) {
+        if (BMTournamentState::END_TOURNAMENT == $tournament->tournamentState) {
+            $status = 'COMPLETE';
+        } elseif (BMTournamentState::CANCELLED == $tournament->tournamentState) {
+            $status = 'CANCELLED';
+        } elseif (BMTournamentState::JOIN_TOURNAMENT == $tournament->tournamentState) {
+            $status = 'OPEN';
+        } else {
+            $status = 'ACTIVE';
+        }
+
+        return $status;
+    }
+
+    /**
+     * Perform a user action on a tournament
+     *
+     * @param int $userId
+     * @param int $tournamentId
+     * @param string $action
+     * @param array $buttonNameArray
+     * @return bool
+     */
+    public function act_on_tournament($userId, $tournamentId, $action, $buttonNameArray = NULL) {
+        switch ($action) {
+            case 'join':
+                return $this->join_tournament($userId, $tournamentId, $buttonNameArray);
+            case 'leave':
+                return $this->leave_tournament($userId, $tournamentId);
+            case 'cancel':
+                return $this->cancel_tournament($userId, $tournamentId);
+            default:
+                error_log('Invalid action on tournament');
```

Comment:
Don't `error_log()` this --- it's not an error.  It's something the user did wrong.  Just set the message and return, we don't want to spam the error log.

Reply:
error_log removed